File: com/tm/java/common/service/monitor/MMonitor.java
package com.tm.java.common.service.monitor;

import dagger.Module;
import dagger.Provides;
import dagger.multibindings.IntoMap;
import io.vertx.core.Verticle;

import com.tm.java.common.annotation.VerticleMapKey;

@Module
public class MMonitor {
    @Provides
    @IntoMap
    @VerticleMapKey(MonitoringVerticle.class)
    Verticle provideMonitoringVerticle(MonitoringVerticle vertical) {
        return vertical;
    }
}

File: com/tm/kotlin/common/mods/monitor/MMonitor.kt
package com.tm.kotlin.common.mods.monitor

import com.tm.kotlin.common.http.API
import com.tm.kotlin.common.http.HttpFactory
import com.tm.kotlin.common.http.MHttpFactory
import com.tm.kotlin.common.mods.base.MBase
import dagger.Module
import dagger.Provides
import dagger.multibindings.IntoMap
import dagger.multibindings.StringKey
import io.micrometer.prometheusmetrics.PrometheusConfig
import io.micrometer.prometheusmetrics.PrometheusMeterRegistry
import io.vertx.core.Verticle
import io.vertx.core.http.HttpMethod
import javax.inject.Singleton

@Module(
    includes = [
        MBase::class,
        MHttpFactory::class
    ]
)
class MMonitor {
    @Provides
    @Singleton
    fun providePrometheusMeterRegistry(): PrometheusMeterRegistry {
        return PrometheusMeterRegistry(PrometheusConfig.DEFAULT)
    }

    @Provides
    @IntoMap
    @StringKey(value = "Monitor")
    fun provideHelloServiceVerticle(
        httpFactory: HttpFactory,
        registry: PrometheusMeterRegistry
    ): Verticle {
        return httpFactory.create(
            "Monitor",
            8089,
            listOf(
                API(
                    HttpMethod.GET,
                    "/metrics"
                ) { ctx ->
                    ctx.response()
                        .putHeader("content-type", "text/plain")
                        .end(registry.scrape())
                }
            )
        )
    }
}